src/game/Maps/MapUpdater.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * This file is part of the CMaNGOS Project. See AUTHORS file for Copyright information
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include "MapUpdater.h"
+#include "MapWorkers.h"
+
+MapUpdater::MapUpdater(size_t num_threads) : _cancelationToken(false), pending_requests(0)
+{
+    for (size_t i = 0; i < num_threads; ++i)
+        _workerThreads.push_back(std::thread(&MapUpdater::WorkerThread, this));
+}
+
+void MapUpdater::activate(size_t num_threads)
+{
+    if (activated())
+        return;
+
+    for (size_t i = 0; i < num_threads; ++i)
+        _workerThreads.push_back(std::thread(&MapUpdater::WorkerThread, this));
+}
+
+void MapUpdater::deactivate()
+{
+    _cancelationToken = true;
+
```

Comment:
By setting the cancelationToken, your worker threads won't be executing any new updates any more and will exit after having finished any ongoing update, if any, so the ```wait()``` below seems unnecessary. I think you might as well just _queue.Cancel() and then join the worker threads, which will wait for them to exit anyway.

This also happens to avoid a double ```wait()``` in ```MapUpdater::join()``` below (on directly, one via ```deactivate()```.